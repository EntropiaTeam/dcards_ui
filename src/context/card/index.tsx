import React, {
  createContext, useContext, useReducer, FC
} from 'react';
import { CardActionsType, UseCardContext } from './types/CardTypes';
import CardReducer, { initialState } from './reducer/CardReducer';
import { CardActions } from './actions/CardActions';

// Create Context
export const CardContext = createContext(initialState);
export const CardActionsContext = createContext({} as CardActionsType);
export const useCardContext = (): UseCardContext => (
  { state: useContext(CardContext), actions: useContext(CardActionsContext) }
);

export const CardProvider: FC = ({ children }) => {
  const [state, dispatch] = useReducer(CardReducer, initialState);

  // Actions
  const actions: CardActionsType = {
    setCardToEdit: CardActions.setCardToEdit(dispatch)
  };

  return (
    <CardContext.Provider value={state}>
      <CardActionsContext.Provider value={actions}>
        {children}
      </CardActionsContext.Provider>
    </CardContext.Provider>
  );
};
