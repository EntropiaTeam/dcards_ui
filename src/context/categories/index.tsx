import React, {
  createContext, useContext, useReducer, useMemo, FC
} from 'react';
import { CategoriesActionsType, UseCategoriesContext } from './types/CategoriesTypes';
import CategoriesReducer, { initialState } from './reducer/CategoriesReducer';
import { CategoriesActions } from './actions/CategoriesActions';

// Create Context
export const CategoriesContext = createContext(initialState);
export const CategoriesActionsContext = createContext({} as CategoriesActionsType);
export const useCategoriesContext = (): UseCategoriesContext => ({
  state: useContext(CategoriesContext), actions: useContext(CategoriesActionsContext)
});

export const CategoriesProvider: FC = ({ children }) => {
  const [state, dispatch] = useReducer(CategoriesReducer, initialState);

  // Actions
  const actions: CategoriesActionsType = useMemo(() => ({
    setCategories: CategoriesActions.setCategories(dispatch)
  }), []);

  return (
    <CategoriesContext.Provider value={state}>
      <CategoriesActionsContext.Provider value={actions}>
        {children}
      </CategoriesActionsContext.Provider>
    </CategoriesContext.Provider>
  );
};
