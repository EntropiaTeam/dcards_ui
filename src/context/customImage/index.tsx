import React, {
  createContext, useContext, useReducer, useMemo, FC
} from 'react';
import { CustomImageActionsType, UseCustomImageContext } from './types/CustomImageTypes';
import { CustomImageActions } from './actions/CustomImageActions';
import CustomImageReducer, { initialState } from './reducer/CustomImageReducer';

// Create Context
export const CustomImageContext = createContext(initialState);
export const CustomImageActionsContext = createContext({} as CustomImageActionsType);
export const useCustomImageContext = (): UseCustomImageContext => (
  { state: useContext(CustomImageContext), actions: useContext(CustomImageActionsContext) }
);

export const CustomImageProvider: FC = ({ children }) => {
  const [state, dispatch] = useReducer(CustomImageReducer, initialState);
  // Actions
  const actions: CustomImageActionsType = useMemo(() => ({
    setErrors: CustomImageActions.setErrors(dispatch),
    setUserImage: CustomImageActions.setUserImage(dispatch),
    setCropper: CustomImageActions.setCropper(dispatch),
    setAgreement: CustomImageActions.setAgreement(dispatch),
    setAgreementTranslation: CustomImageActions.setAgreementTranslation(dispatch),
    setCroppedImage: CustomImageActions.setCroppedImageValue(dispatch),
    setCropperBackImageURL: CustomImageActions.setCropperBackImageValue(dispatch),
    setRotationDegree: CustomImageActions.setRotationDegree(dispatch),
    setZoomRatio: CustomImageActions.setZoomRatio(dispatch),
    setCroppedCanvasData: CustomImageActions.setCroppedCanvasData(dispatch),
    resetCustomImageState: CustomImageActions.resetCustomImageState(dispatch)
  }), [dispatch]);

  return (
    <CustomImageContext.Provider value={state}>
      <CustomImageActionsContext.Provider value={actions}>
        {children}
      </CustomImageActionsContext.Provider>
    </CustomImageContext.Provider>
  );
};
