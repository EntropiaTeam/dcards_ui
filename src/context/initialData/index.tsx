import React, {
  createContext, useContext, useReducer, useMemo, FC
} from 'react';
import { UseInitialDataContext, InitialDataActionsType } from './types/InitialDataTypes';
import InitialDataReducer, { initialState as initDatainitialState } from './reducer/InitialDataReducer';
import { InitialDataActions } from './actions/InitialDataActions';

import {
  useCategoriesContext,
  useCustomImageContext
} from '..';

// Create Context
export const InitialDataContext = createContext(initDatainitialState);
export const InitialDataActionsContext = createContext({} as InitialDataActionsType);
export const useInitialDataContext = (): UseInitialDataContext => ({
  state: useContext(InitialDataContext), actions: useContext(InitialDataActionsContext)
});

export const InitialDataProvider: FC = ({ children }) => {
  const { actions: { setCategories } } = useCategoriesContext();
  const { actions: { setAgreementTranslation } } = useCustomImageContext();
  const [initDatastate, dispatchInitData] = useReducer(InitialDataReducer, initDatainitialState);

  // Actions
  const actions: InitialDataActionsType = useMemo(() => ({
    loadInitialData: InitialDataActions.loadInitialData(
      dispatchInitData,
      setCategories,
      setAgreementTranslation
    )
  }), [setCategories, setAgreementTranslation]);

  return (
    <InitialDataContext.Provider value={initDatastate}>
      <InitialDataActionsContext.Provider value={actions}>
        {children}
      </InitialDataActionsContext.Provider>
    </InitialDataContext.Provider>
  );
};
