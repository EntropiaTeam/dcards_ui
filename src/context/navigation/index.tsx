import React, {
  createContext, useContext, useReducer, FC
} from 'react';
import { NavigationActionTypes, UseNavigationContext } from './types/NavigationTypes';
import NavigationReducer, { initialState } from './reducer/NavigationReducer';
import { NavigationActions } from './actions/NavigationActions';

// Create Context
export const NavigationContext = createContext(initialState);
const NavigationActionsContext = createContext({} as NavigationActionTypes);
export const useNavigationContext = (): UseNavigationContext => (
  { state: useContext(NavigationContext), actions: useContext(NavigationActionsContext) }
);

export const NavigationProvider: FC = ({ children }) => {
  const [state, dispatch] = useReducer(NavigationReducer, initialState);

  // Actions
  const actions: NavigationActionTypes = {
    changeIsBackButtonVisible: NavigationActions.changeIsBackButtonVisible(dispatch),
    changeIframeParams: NavigationActions.changeIframeParams(dispatch)
  };

  return (
    <NavigationContext.Provider value={state}>
      <NavigationActionsContext.Provider value={actions}>
        {children}
      </NavigationActionsContext.Provider>
    </NavigationContext.Provider>
  );
};
