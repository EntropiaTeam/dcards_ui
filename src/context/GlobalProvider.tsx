import React from 'react';
import {
  CardProvider,
  CategoriesProvider,
  TextProvider,
  OrderProvider,
  CustomImageProvider
} from '.';
import { NavigationProvider } from './navigation';

export const GlobalProvider: React.FunctionComponent = ({ children }) => (
  <NavigationProvider>
    <CardProvider>
      <CategoriesProvider>
        <TextProvider>
          <CustomImageProvider>
            <OrderProvider>
              {children}
            </OrderProvider>
          </CustomImageProvider>
        </TextProvider>
      </CategoriesProvider>
    </CardProvider>
  </NavigationProvider>
);
