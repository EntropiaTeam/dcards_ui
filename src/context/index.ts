import { CardProvider, useCardContext } from './card';
import { CategoriesProvider, useCategoriesContext } from './categories';
import { TextProvider, useTextContext } from './text';
import { OrderProvider, useOrderContext } from './order';
import { CustomImageProvider, useCustomImageContext } from './customImage';
import { NavigationProvider, useNavigationContext } from './navigation';

export {
  CardProvider,
  useCardContext,
  CategoriesProvider,
  useCategoriesContext,
  TextProvider,
  useTextContext,
  OrderProvider,
  useOrderContext,
  CustomImageProvider,
  useCustomImageContext,
  NavigationProvider,
  useNavigationContext
};
