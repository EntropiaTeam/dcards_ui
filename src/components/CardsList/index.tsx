import React, { memo, FC } from 'react';
import { Grid } from '@material-ui/core';
import { Theme, makeStyles } from '@material-ui/core/styles';
import { useHistory, useRouteMatch } from 'react-router-dom';
import CardImage from '../UI/CardImage';
import ImageContainer from '../UI/ImageContainer';
import { useCardContext } from '../../context';
import { SingleCard } from '../../context/card/types/CardTypes';
import useAppConfig from '../../hooks/useAppConfig';
import CardUtils from '../../utils/Card';
import { RoutePath } from '../../enums/Routes';
import { getImageUrl } from '../../helpers';

const useStyles = makeStyles((theme: Theme) => ({
  root: {
    display: 'flex',
    flexWrap: 'wrap',
    flexDirection: 'row',
    backgroundColor: theme.palette.background.paper
  },
  cardContainer: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    marginBottom: 30,
    padding: '0 10px',
    [theme.breakpoints.only('xs')]: {
      marginBottom: 20
    }
  },
  imgContainer: {
    border: '1px solid rgba(0, 0, 0, 0.1)',
    boxShadow: '0px 7px 49px rgba(202, 204, 213, 0.45)'
  }
}));

type Props = {
  cards: SingleCard[];
};

const CardsList: FC<Props> = ({ cards }) => {
  const classes = useStyles();
  const history = useHistory();
  const catalogRouteMatch = useRouteMatch({ path: RoutePath.Catalog, exact: true });
  const { appConfig } = useAppConfig();
  const { actions: { setCardToEdit } } = useCardContext();

  const handleClick = (card: SingleCard) => (): void => {
    const targetLocation = `${catalogRouteMatch!.url}/${card.id}/edit/photo`;
    setCardToEdit(card);

    if (CardUtils.isCustomCard(card)) {
      history.push({ pathname: `${catalogRouteMatch!.url}/agreement`, state: { targetLocation } });
      return;
    }
    if (catalogRouteMatch) history.push(targetLocation);
  };

  const handleCardKeyDown = (
    card: SingleCard,
    event: React.KeyboardEvent<HTMLDivElement>
  ): void => {
    if (event.key === 'Enter') {
      handleClick(card)();
    }
  };

  const cardTabIndexIsset = 3;

  const renderCards = cards.length && cards.map((card: SingleCard, idx: number) => (
    <Grid
      item
      xs={6}
      sm={6}
      md={4}
      lg={4}
      xl={4}
      key={`card-${card.category[0]}-${card.id}`}
      className={classes.cardContainer}
      onKeyDown={(event) => {
        handleCardKeyDown(card, event);
      }}
    >
      <ImageContainer
        imgConfig={appConfig.cardsListImgConfig}
        cls={classes.imgContainer}
        onClick={handleClick(card)}
        tabIndex={idx + cardTabIndexIsset}
      >
        <CardImage
          imageUrl={getImageUrl(`${card.id}_front.png`, 'gallery')}
          cardName={card.name}
        />
      </ImageContainer>
    </Grid>
  ));

  return (
    <Grid container className={classes.root}>
      {renderCards}
    </Grid>
  );
};

export default memo(CardsList);
