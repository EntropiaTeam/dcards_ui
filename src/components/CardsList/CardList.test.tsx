import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import CardsList from '.';
import { CardProvider } from '../../context/card';

const mockHistoryPush = jest.fn();

interface ReactRouterDom {
  useHistory: () => void;
  useRouteMatch: () => void;
}

jest.mock('react-router-dom', (): ReactRouterDom => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    push: mockHistoryPush
  }),
  useRouteMatch: jest.fn().mockReturnValue(true)
}));

describe('<CardsList />', () => {
  const props = {
    cards: [{
      category: 'Test',
      id: 0,
      attributes: ['Photo']
    },
    {
      category: 'Test2',
      id: 1,
      attributes: ['Photo']
    }]
  };

  const CardsListComponent = renderer.create(
    <CardProvider>
      <CardsList {...props} />
    </CardProvider>
  ).toJSON();

  const wrapper = mount(
    <CardProvider>
      <CardsList {...props} />
    </CardProvider>
  );
  const firstCard = wrapper.find('.MuiGrid-item');

  it('should track history change on Enter pressed', () => {
    act(() => {
      firstCard.first().props().onKeyDown({ key: 'Enter' });
    });
    expect(mockHistoryPush).toHaveBeenCalled();
  });

  it('should render correct number cards', () => {
    expect(firstCard).toHaveLength(props.cards.length);
  });

  it('should render CardsList component correctly', () => {
    expect(CardsListComponent).toMatchSnapshot();
  });

  it('should render CardsList component with proper cards prop', () => {
    expect((wrapper).childAt(0).prop('cards')).toEqual(props.cards);
  });
});
