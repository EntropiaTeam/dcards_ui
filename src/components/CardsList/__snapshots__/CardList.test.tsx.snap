// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardsList /> should render CardsList component correctly 1`] = `
<div
  className="MuiGrid-root makeStyles-root-1 MuiGrid-container"
>
  <div
    className="MuiGrid-root makeStyles-cardContainer-2 MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-4 MuiGrid-grid-xl-4"
    onKeyDown={[Function]}
  >
    <div
      className="MuiBox-root MuiBox-root-6 makeStyles-imageContainer-4 makeStyles-imageContainer-5 makeStyles-imgContainer-3"
      onClick={[Function]}
      tabIndex={3}
    >
      <img
        className="makeStyles-image-7 makeStyles-image-8"
        src="/assets/i/0_gallery_front.png"
      />
    </div>
  </div>
  <div
    className="MuiGrid-root makeStyles-cardContainer-2 MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-4 MuiGrid-grid-xl-4"
    onKeyDown={[Function]}
  >
    <div
      className="MuiBox-root MuiBox-root-10 makeStyles-imageContainer-4 makeStyles-imageContainer-9 makeStyles-imgContainer-3"
      onClick={[Function]}
      tabIndex={4}
    >
      <img
        className="makeStyles-image-7 makeStyles-image-11"
        src="/assets/i/1_gallery_front.png"
      />
    </div>
  </div>
</div>
`;
